### PR TITLE
[WIP] Upgrade to Pydantic v2

### PR DESCRIPTION
**Pull request by Leobouloc**
## Purpose

As mentionned in https://github.com/openfun/ralph/issues/401 , upgrading to pydantic 2 should greatly improve performances. Also it would allow projects depending on `ralph` (such as `warren`) to upgrade more easily. 

### ERROR: Extra not permitted

In `conf.py`, errors such as: 
```
    super().__init__(
E   pydantic_core._pydantic_core.ValidationError: 1 validation error for Settings
E   LOCALE_ENCODING
E     Extra inputs are not permitted [type=extra_forbidden, input_value='utf-8', input_type=str]
E       For further information visit https://errors.pydantic.dev/2.1/v/extra_forbidden
```

may be due to a change in default behavior in models `extra=ignore` to `extra=failure`. We may set this parameter manually to fix this issue.

### ERROR: Unable to generate pydantic-core schema 
```
E   pydantic.errors.PydanticSchemaGenerationError: Unable to generate pydantic-core schema for <class 'ralph.models.xapi.base.common.IRI'>. Set `arbitrary_types_allowed=True` in the model_config to ignore this error or implement `__get_pydantic_core_schema__` on your type to fully support it.
E
E   If you got this error by calling handler(<some type>) within `__get_pydantic_core_schema__` then you likely need to call `handler.generate_schema(<some type>)` since we do not call `__get_pydantic_core_schema__` on `<some type>` otherwise to avoid infinite recursion.
E
```
To fix this we change the way we define classes.

**Previously in Ralph**
```
class CommaSeparatedTuple(BaseModel):
    @model_validator(mode='before')
    @classmethod
    def validate(cls, value: Union[str, Tuple[str]]) -> Tuple[str]:
        """Checks whether the value is a comma separated string or a tuple."""

        if isinstance(value, tuple):
            return value

        if isinstance(value, str):
            return tuple(value.split(","))

        raise TypeError("Invalid comma separated list")
```
**New in Ralph** 
```
def validate_comma_separated_tuple(value: Union[str, Tuple[str, ...]]) -> Tuple[str]:
    """Checks whether the value is a comma separated string or a tuple."""

    if isinstance(value, tuple):
        return value

    if isinstance(value, str):
        return tuple(value.split(","))

    raise TypeError("Invalid comma separated list")

CommaSeparatedTuple = Annotated[Union[str, Tuple[str, ...]], AfterValidator(validate_comma_separated_tuple)]
```

### ERROR: Missing values

A recurrent error is the one below, which should be linked to the change of behavior when using `Optional[...]` in Pydantic V2. Previously, it would assign a default value of `None`, which is no longer the case. This should have been dealt with by `bump-pydantic` but was not always the case. It would be nice to know why before confirming this solution.
```
src/ralph/models/xapi/video/statements.py:85: in VideoPlayed
    verb: PlayedVerb = PlayedVerb()
E   pydantic_core._pydantic_core.ValidationError: 1 validation error for PlayedVerb
E   display
E     Field required [type=missing, input_value={}, input_type=dict]
E       For further information visit https://errors.pydantic.dev/2.1/v/missing
```

**Previously in Ralph**
```
class VideoActivity(BaseXapiActivity):
    name: Optional[Dict[Literal[LANG_EN_US_DISPLAY], str]]
```
**New in Ralph** 
```
class VideoActivity(BaseXapiActivity):
    name: Optional[Dict[Literal[LANG_EN_US_DISPLAY], str]] = None
```

### CODE REVIEW

File: src/ralph/conf.py [hunk 1]
```diff
@@ -124,11 +131,11 @@ class MongoClientOptions(ClientOptions):
 
 
 class ESDatabaseBackendSettings(InstantiableSettingsItem):
-    """Pydantic modelf for Elasticsearch database backend configuration settings."""
+    """Pydantic modelf for Elasticsearch database backend configuration settings."""
```
Comment (quitterie-lcs) [inline src/ralph/conf.py:134]:
```suggestion
    """Pydantic model for Elasticsearch database backend configuration settings."""    
```

File: src/ralph/conf.py [hunk 2]
```diff
@@ -3,12 +3,12 @@
 import io
 from enum import Enum
 from pathlib import Path
-from typing import List, Tuple, Union
+from typing import Any, List, Tuple, Union
```
Comment (quitterie-lcs) [inline src/ralph/conf.py:6]:
```suggestion
from typing import Any, List, Optional, Tuple, Union
```

File: src/ralph/conf.py [hunk 3]
```diff
@@ -3,12 +3,12 @@
...
 
 try:
-    from typing import Literal
+    from typing import Annotated, Literal, Optional
```
Comment (quitterie-lcs) [inline src/ralph/conf.py:9]:
```suggestion
    from typing import Annotated, Literal
```

File: src/ralph/conf.py [hunk 4]
```diff
@@ -3,12 +3,12 @@
...
 except ImportError:
-    from typing_extensions import Literal
+    from typing_extensions import Annotated, Literal, Optional
```
Comment (quitterie-lcs) [inline src/ralph/conf.py:11]:
```suggestion
    from typing_extensions import Annotated, Literal
```

File: src/ralph/models/edx/navigational/statements.py [hunk 5]
```diff
@@ -74,7 +74,8 @@ class UISeqNext(BaseBrowserModel):
     event_type: Literal["seq_next"]
     name: Literal["seq_next"]
 
-    @validator("event")
+    @field_validator("event")
+    @classmethod
```
Comment (quitterie-lcs) [inline src/ralph/models/edx/navigational/statements.py:78]:
```suggestion
```

File: src/ralph/models/edx/navigational/statements.py [hunk 6]
```diff
@@ -104,7 +105,8 @@ class UISeqPrev(BaseBrowserModel):
     event_type: Literal["seq_prev"]
     name: Literal["seq_prev"]
 
-    @validator("event")
+    @field_validator("event")
+    @classmethod
```
Comment (quitterie-lcs) [inline src/ralph/models/edx/navigational/statements.py:109]:
```suggestion
```

File: src/ralph/models/edx/open_response_assessment/fields/events.py [hunk 7]
```diff
@@ -4,13 +4,13 @@
 from typing import Dict, List, Optional, Union
 
 try:
-    from typing import Literal
+    from typing import Annotated, iteral
```
Comment (quitterie-lcs) [inline src/ralph/models/edx/open_response_assessment/fields/events.py:7]:
```suggestion
    from typing import Annotated, Literal
```

File: src/ralph/models/edx/problem_interaction/fields/events.py [hunk 8]
```diff
@@ -40,13 +40,13 @@ class CorrectMap(BaseModelWithConfig):
         queuestate (json): see QueueStateField.
     """
 
-    answervariable: Union[Literal[None], None, str]
+    answervariable: Union[Literal[None], None, str] = None
```
Comment (quitterie-lcs) [inline src/ralph/models/edx/problem_interaction/fields/events.py:43]: I wonder if setting a default value for this field makes it not required

File: src/ralph/models/edx/textbook_interaction/fields/events.py [hunk 9]
```diff
@@ -262,16 +262,16 @@ class BookEventField(AbstractBaseEventField):
...
             r"^\/asset-v1:[^\/+]+(\/|\+)[^\/+]+(\/|\+)[^\/?]+type@asset\+block.+$"  # noqa
         )
-    )
+    )]
     name: Union[
         Literal["textbook.pdf.page.loaded"], Literal["textbook.pdf.page.navigatednext"]
     ]
     new: int
-    old: Optional[int]
+    old: Optional[int] = None # TODO: validate that this is the behavior we want
```
Comment (quitterie-lcs) [inline src/ralph/models/edx/textbook_interaction/fields/events.py:274]: Why this TODO for this field in particular?

File: src/ralph/models/xapi/base/common.py [hunk 10]
```diff
@@ -1,52 +1,36 @@
...
-    def __get_validators__(cls):  # noqa: D105
-        def validate(mailto: str):
-            """Checks whether the provided value follows the `mailto:email` format."""
-            if not mailto.startswith("mailto:"):
-                raise TypeError("Invalid `mailto:email` value")
-            valid = validate_email(mailto[7:])
-            return cls(f"mailto:{valid[1]}")
-
-        yield validate
+MailtoEmail = Annotated[str, AfterValidator(validate_mailto_email)]
```
Comment (quitterie-lcs) [inline src/ralph/models/xapi/base/common.py:36]:
```suggestion
MailtoEmail = Annotated[str, AfterValidator(validate_mailto_email)]

```

File: src/ralph/models/xapi/concepts/activity_types/video.py [hunk 11]
```diff
@@ -37,5 +37,5 @@ class VideoActivity(BaseXapiActivity):
         definition (dict): See VideoActivityDefinition.
     """
 
-    name: Optional[Dict[Literal[LANG_EN_US_DISPLAY], str]]
+    name: Optional[Dict[Literal[LANG_EN_US_DISPLAY], str]] = None # TODO: validate that this is the behavior we want
```
Comment (quitterie-lcs) [inline src/ralph/models/xapi/concepts/activity_types/video.py:40]: Ditto and for the rest of the fields you have commented this way

**Review by quitterie-lcs — COMMENTED** [review]
Great job!

File: setup.cfg [hunk 12]
```diff
@@ -30,7 +30,8 @@ install_requires =
     ; By default, we only consider core dependencies required to use Ralph as a
     ; library (mostly models).
     langcodes>=3.2.0
-    pydantic[dotenv,email]>=1.10.0, <2.0
+    pydantic[email]>=2.1.1, <3.0
```
Comment (quitterie-lcs) [inline setup.cfg:33]:
```suggestion
    pydantic[email]>=2.2.0
```

**Comment by Leobouloc** [comment]
replaced by https://github.com/openfun/ralph/pull/504

File: src/ralph/models/edx/problem_interaction/fields/events.py (hunk 8 above)
Comment (Leobouloc) [reply src/ralph/models/edx/problem_interaction/fields/events.py:43]: comment transfered to newer PR

File: src/ralph/models/edx/navigational/statements.py (hunk 5 above)
Comment (Leobouloc) [reply src/ralph/models/edx/navigational/statements.py:78]: removed

File: src/ralph/models/edx/navigational/statements.py (hunk 6 above)
Comment (Leobouloc) [reply src/ralph/models/edx/navigational/statements.py:109]: removed

**Pull request closed without merging**